File: Arrays/Easy/removeDuplicates.cpp
#include<iostream>
#include<vector>
#include<set>

using namespace std;


void removeDup_brute(vector <int>&arr,int n){

    set<int> map;

    for(int i=0;i<n;i++){

        map.insert(arr[i]);   //O(n log n)  
    }

    int index=0;

    for(auto &i:map){
        arr[index]=i;           //O(n)
        index++;
    }

    
    for(int i=0;i<index;i++) cout<<arr[i] << " ";
    
    for(int i=index;i<n;i++) arr[i] =-1;
    
    cout<<endl;
    for(int i=0;i<n;i++) cout<<arr[i] << " ";
}


void removeDup_optimize(vector<int> &arr,int n){

    int pointer=0;
    

    for(int i=1;i<n;i++){

        if(arr[i]!=arr[pointer]){

            arr[pointer+1] = arr[i];
            pointer++;

        }
    }

   
    for(int i=0;i<=pointer;i++){
        cout<<arr[i]<<" ";
    }
}

int main(){

    vector<int> arr={1,1,2,2,3,4,5,6,7,7};
    int n=arr.size();

    // removeDup_brute(arr,n);
    removeDup_optimize(arr,n);
    return 0;
}

File: Arrays/Python/Easy/removeDuplicates.py


def removeDup_brute(nums,n):

    unique=set()          #O(n logn)

    for num in nums:      #O(n)
        unique.add(num)

    
    index=0
    for unum in unique:         #~~ O(n-x)
        nums[index] = unum
        index+=1
    
  

    for i in range(index,n):     # ~O(x)
        nums[i]='_'

    return nums


def removeDup_Optimal(nums,n):

    pointer=0

    for i in range(1,n):

        if nums[i] != nums[pointer]:
            nums[pointer+1]=nums[i]
            pointer+=1

    return nums[:pointer+1]


nums=[1,1,2,2,3,4,5,6,7,7]
n=len(nums)

# print(removeDup_brute(nums,n))   # O(nlog) + O(n)
print(removeDup_Optimal(nums,n))   #O(n)